### PR TITLE
[CLI] add domain migration validator command

### PR DESCRIPTION
**Pull request by manasaling**
**What changed?**
Added meta data validation checker, long workflow checker, search attributes checker and dynamic config checker for domain portability.

**Why?**
All these checks are necessary to validate whether or not a user can port a certain domain. By having these checks, we can reduce the time spent on validation for domain portability.

**How did you test it?**

Search Attributes Valid Case:
Screenshot 2023-08-01 at 9 16 42 AM

Search Attributes False Case:
Screenshot 2023-08-01 at 9 17 52 AM

More Testing: https://docs.google.com/spreadsheets/d/1JJRWRbWrU3G1NBGFSWr89wWbsex8xfX_uRlAFqIiMss/edit#gid=0 

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: config/dynamicconfig/development_es.yaml [hunk 1]
```diff
@@ -42,5 +42,4 @@ system.minRetentionDays:
     - value: 0
 history.EnableConsistentQueryByDomain:
 - value: true
-  constraints: {}
-
+  constraints: {}
```
Comment (shijiesheng) [inline config/dynamicconfig/development_es.yaml:45]: nit: unnecessary change, keep the line

File: tools/cli/domainCommands.go [hunk 2]
```diff
@@ -681,6 +773,398 @@ func (d *domainCLIImpl) describeDomain(
...
+
+	currResp, err := d.frontendClient.DescribeDomain(ctx, request)
+
+	if err != nil {
+		ErrorAndExit(fmt.Sprintf("Could not describe old domain, Please check to see if old domain exists before migrating."), err)
+	}
+	newResp, err := d.destinationClient.DescribeDomain(ctx, request2)
+	if err != nil {
+		ErrorAndExit(fmt.Sprintf("Could not describe new domain, Please check to see if new domain exists before migrating."), err)
+	}
```
Comment (shijiesheng) [inline tools/cli/domainCommands.go:830]: nit: inline initialization of entities that are used only once
```suggestion
	ctx, cancel := newContext(c)
	defer cancel()

	currResp, err := d.frontendClient.DescribeDomain(ctx, &types.DescribeDomainRequest{
		Name: &domain,
	})
	if err != nil {
		ErrorAndExit(fmt.Sprintf("Could not describe old domain, Please check to see if old domain exists before migrating."), err)
	}
	newResp, err := d.destinationClient.DescribeDomain(ctx, &types.DescribeDomainRequest{
		Name: &newDomain,
	})
	if err != nil {
		ErrorAndExit(fmt.Sprintf("Could not describe new domain, Please check to see if new domain exists before migrating."), err)
	}
```

File: tools/cli/domainCommands.go [hunk 3]
```diff
@@ -681,6 +773,398 @@ func (d *domainCLIImpl) describeDomain(
...
+		ValidationResult: validationResult,
+		ValidationDetails: ValidationDetails{
+			CurrentDomainRow: currResp,
+			NewDomainRow:     newResp,
+		},
+	}
+	return validationRow
+}
+
+func metaDataValidation(currResp *types.DescribeDomainResponse, newResp *types.DescribeDomainResponse) bool {
```
Comment (shijiesheng) [inline tools/cli/domainCommands.go:843]: return validation result (bool) as well as mismatched message (string)

File: tools/cli/domainCommands.go [hunk 4]
```diff
@@ -681,6 +773,398 @@ func (d *domainCLIImpl) describeDomain(
...
+	newResp, err := d.destinationClient.DescribeDomain(ctx, request2)
+	if err != nil {
+		ErrorAndExit(fmt.Sprintf("Could not describe new domain, Please check to see if new domain exists before migrating."), err)
+	}
+	validationResult := metaDataValidation(currResp, newResp)
+	validationRow := DomainMigrationRow{
+		ValidationCheck:  "Domain Meta Data",
+		ValidationResult: validationResult,
+		ValidationDetails: ValidationDetails{
+			CurrentDomainRow: currResp,
```
Comment (shijiesheng) [inline tools/cli/domainCommands.go:836]: add a new string field `MismatchedDomainMetaData` and render it if len>0. get the string from `metaDataValidation` output.

File: tools/cli/domainCommands.go [hunk 5]
```diff
@@ -681,6 +773,398 @@ func (d *domainCLIImpl) describeDomain(
...
+			NewDomainRow:     newResp,
+		},
+	}
+	return validationRow
+}
+
+func metaDataValidation(currResp *types.DescribeDomainResponse, newResp *types.DescribeDomainResponse) bool {
+	if !reflect.DeepEqual(currResp.Configuration, newResp.Configuration) {
+		return false
```
Comment (shijiesheng) [inline tools/cli/domainCommands.go:845]:
```suggestion
		return false, "mismatched DomainConfiguration"
```

File: tools/cli/domainCommands.go [hunk 6]
```diff
@@ -681,6 +773,398 @@ func (d *domainCLIImpl) describeDomain(
...
+	}
+
+	if currResp.DomainInfo.OwnerEmail != newResp.DomainInfo.OwnerEmail {
+		return false
```
Comment (shijiesheng) [inline tools/cli/domainCommands.go:849]:
```suggestion
		return false, "mismatched OwnerEmail"
```

File: tools/cli/domainCommands.go [hunk 7]
```diff
@@ -681,6 +773,398 @@ func (d *domainCLIImpl) describeDomain(
...
+	}
+	return true
```
Comment (shijiesheng) [inline tools/cli/domainCommands.go:851]:
```suggestion
	return true, ""
```

File: tools/cli/domainCommands.go [hunk 8]
```diff
@@ -681,6 +773,398 @@ func (d *domainCLIImpl) describeDomain(
...
+	if countWorkFlows > 0 {
+		ErrorAndExit(fmt.Sprintf("Number of workflows is > 0"), nil)
+	}
+	return DomainMigrationRow{
+		ValidationCheck:  "Workflow Check",
+		ValidationResult: true,
+		ValidationDetails: ValidationDetails{
+			LongRunningWorkFlowNum: &countWorkFlows,
+		},
+	}
```
Comment (shijiesheng) [inline tools/cli/domainCommands.go:868]: Instead of ErrorAndExit, maybe return false check
```suggestion
        check := countWorkFlows == 0
	return DomainMigrationRow{
		ValidationCheck:  "Workflow Check",
		ValidationResult: check,
		ValidationDetails: ValidationDetails{
			LongRunningWorkFlowNum: &countWorkFlows,
		},
	}
```

File: tools/cli/domainCommands.go [hunk 9]
```diff
@@ -681,6 +773,398 @@ func (d *domainCLIImpl) describeDomain(
...
+				dynamicconfig.DomainFilter(currDomain),
+			})
+
+			newRequest := dynamicconfig.ToGetDynamicConfigFilterRequest(configKey.String(), []dynamicconfig.FilterOption{
+				dynamicconfig.DomainFilter(newDomain),
+			})
+
+			currResp, err := d.frontendAdminClient.GetDynamicConfig(ctx, currRequest)
+			if err != nil {
+				currResp = &types.GetDynamicConfigResponse{}
```
Comment (shijiesheng) [inline tools/cli/domainCommands.go:995]: add a comment here
```suggestion
                                  // empty to indicate N/A
				currResp = &types.GetDynamicConfigResponse{}
```

File: tools/cli/domainUtils.go [hunk 10]
```diff
@@ -201,6 +203,30 @@ var (
 		getFormatFlag(),
 	}
 
+	migrateDomainFlags = []cli.Flag{
+
+		cli.StringFlag{
+			Name:  FlagDestinationAddress,
+			Usage: "t",
```
Comment (shijiesheng) [inline tools/cli/domainUtils.go:210]: add usage `destination cadence-frontend address in <host>:<port> format`

File: tools/cli/domainUtils.go [hunk 11]
```diff
@@ -201,6 +203,30 @@ var (
...
+		},
+		cli.StringFlag{
+			Name:  FlagDestinationDomain,
+			Usage: "t",
```
Comment (shijiesheng) [inline tools/cli/domainUtils.go:214]: `destination domain name`

File: tools/cli/domainUtils.go [hunk 12]
```diff
@@ -201,6 +203,30 @@ var (
...
+		},
+
+		cli.StringSliceFlag{
+			Name:  FlagTaskList,
+			Usage: "t",
```
Comment (shijiesheng) [inline tools/cli/domainUtils.go:219]: `all tasklists in the current domain`

File: tools/cli/factory.go [hunk 13]
```diff
@@ -76,20 +76,28 @@ type ClientFactory interface {
...
+		return c.GlobalString(FlagAddress)
+	})
+}
+
+func newClientFactory(f func(c *cli.Context) string) ClientFactory {
 	logger, err := zap.NewDevelopment()
 	if err != nil {
 		panic(err)
 	}
-
```
Comment (shijiesheng) [inline tools/cli/factory.go:90]: nit: keep the line

File: tools/cli/factory.go [hunk 14]
```diff
@@ -76,20 +76,28 @@ type ClientFactory interface {
...
 	return &clientFactory{
-		logger: logger,
+		addressFlagFunc: f,
+		logger:          logger,
```
Comment (shijiesheng) [inline tools/cli/factory.go:100]: nit: keep the original order

File: tools/cli/factory.go [hunk 15]
```diff
@@ -156,9 +164,10 @@ func (b *clientFactory) ensureDispatcher(c *cli.Context) {
 	if shouldUseGrpc {
 		b.hostPort = grpcPort
 	}
-	if addr := c.GlobalString(FlagAddress); addr != "" {
+	if addr := b.addressFlagFunc(c); addr != "" {
 		b.hostPort = addr
 	}
+
```
Comment (shijiesheng) [inline tools/cli/factory.go:170]: nit: remove the line

File: tools/cli/factory.go [hunk 16]
```diff
@@ -156,9 +164,10 @@ func (b *clientFactory) ensureDispatcher(c *cli.Context) {
 	if shouldUseGrpc {
 		b.hostPort = grpcPort
 	}
-	if addr := c.GlobalString(FlagAddress); addr != "" {
```
Comment (shijiesheng) [inline tools/cli/factory.go:159]: what is this change for?

File: tools/cli/flags.go [hunk 17]
```diff
@@ -297,6 +299,7 @@ const (
 	FlagIsolationGroupSetDrains           = "set-drains"
 	FlagIsolationGroupJSONConfigurations  = "json"
 	FlagIsolationGroupsRemoveAllDrains    = "remove-all-drains"
+	FlagSearchAttribute                   = "search-attr"
```
Comment (shijiesheng) [inline tools/cli/flags.go:302]: nit: seems like all other flags are snake case `search_attr`

File: tools/cli/factory.go (hunk 16 above)
Comment (manasaling) [reply tools/cli/factory.go:159]: i think it's because we created addressFlagFunc func(c *cli.Context) string in the clientFactory struct and we used that instead of c.GlobalString

**Review by shijiesheng — APPROVED** [review]
